src/main/java/rx/schedulers/TrampolineScheduler.java

```diff
@@ -71,14 +71,17 @@ private Subscription enqueue(Action0 action, long execTime) {
                 return Subscriptions.unsubscribed();
             }
             final TimedAction timedAction = new TimedAction(action, execTime, COUNTER_UPDATER.incrementAndGet(TrampolineScheduler.this));
-            queue.add(timedAction);
+            synchronized (queue) {
+                queue.add(timedAction);
+            }
 
             if (wip.getAndIncrement() == 0) {
                 do {
-                    TimedAction polled = queue.poll();
-                    // check for null as it could have been unsubscribed and removed
-                    if (polled != null) {
-                        polled.action.call();
+                    synchronized (queue) {
+                        if (!queue.isEmpty()) {
+                            TimedAction polled = queue.poll();
+                            polled.action.call();
```

Comment:
You should not call the action while holding the lock. Just do a synchronized poll and call the action if you got a non-null value.

In addition, I think the ```COUNTER_UPDATER``` and ```counter``` can be moved into the worker itself because it can ensure total order locally and we can avoid a global serialization point.